src/sa.py

```diff
@@ -7,8 +14,187 @@ def main():
+    '''
+    n = len(x)
+    alpha = alphabet_from_input_string(x)
+    #suffixes = [x[i:]+i*'$' for i, _ in enumerate(x)]
+    x += '$'
+    suffixes = [suffix(x, i) for i, _ in enumerate(x)]
+
+    sort = radix_sort(suffixes, alpha)
+    return [x.i for x in sort]
+
+
+def sa_construction_simple_nsq(x: str) -> list[int]:
+    '''
+    simple sa construction uing builtin strcmp and sort, for comparisson
+    '''
+    suf = [x[i:] for i, _ in enumerate(x)]
+    suf = sorted(suf)
+    return [len(x) - len(i) for i in suf]
+
+
+def lower_bound(a: str, i: int, lo: int, hi: int, x: str, sa: list[int]) -> int:
+    ''' binary search suffix array for lower bound for occurence of character a at position i in a pattern, in string x, given Suffix array SA of x'''
+    while hi > lo:
+        mid = (hi+lo)//2
+        if x[(sa[mid] + i) % len(x)] < a:
+            lo = mid+1
+        else:
+            hi = mid
+    return lo
+
+
+def upper_bound(a: str, i: int, lo: int, hi: int, x: str, sa: list[int]) -> int:
+    ''' binary search suffix array for upper bound for occurence of character a at position i in a pattern, in string x, given Suffix array SA of x'''
+    while hi > lo:
+        mid = (hi+lo)//2
+        if ord(x[(sa[mid] + i) % len(x)]) < ord(a)+1:
+            lo = mid+1
+        else:
+            hi = mid
+    return lo
+
+
+def pattern_match(x: str, p: str, sa: list[int]) -> Iterable[int]:
+    ''' report all occurences of pattern p in string x, given Suffix array: SA of x. '''
+    high = len(sa)
+    low = 0
+
+    if len(p) == 0 or len(x) == 0:
+        return []
+    for i, a in enumerate(p):
+        low = lower_bound(a, i, low, high, x, sa)
+        high = upper_bound(a, i, low, high, x, sa)
+
+    return sa[low:high]
+
+
+# Work in progress below
+def radix_by_index(x: str, u: list[int], alphabet: list[str]) -> dict[str, str]:
+    buckets = {i: 0 for i in alphabet}
+    triplets = get_triplet_map(x, u)
+
+    arr = [i for i in u]  # make copy so we don't override u
+    sa = [0] * len(u)
+    for i in range(3):
+        for trip in arr:
+            buckets[triplets[trip][i]] += 1
+        accsum = 0
+        for bucket in buckets:
+            buckets[bucket], accsum = accsum, accsum + buckets[bucket]
+        for trip in arr:
+            sa[buckets[triplets[trip][i]]] = trip
+            buckets[triplets[trip][i]] += 1
+        for bucket in buckets:
+            buckets[bucket] = 0
+        arr, sa = sa, arr
+
+    i = 0
+    sigma: dict[str, str] = {}
+    for trip in arr:
+        if triplets[trip] not in sigma:
+            sigma[triplets[trip]] = str(i)
+            i += 1
+    sigma = {str(sigma[key]): key for key in sigma}
+    return sigma
+
+
+def get_triplet_map(x: str, u: list[int]) -> dict[int, str]:
+    triplets: dict[int, str] = {}
+    for suf in u:
+        suffix = x[suf:]
+        if len(suffix) >= 3:
+            triplet = suffix[:3]
+        else:
+            triplet = suffix + (3-len(suffix)) * suffix[-1]
```

Comment:
You know that it is the sentinel, so use `'$'` (or `'\0'`). This is dangerous if you at some point allow strings that are not sentinel-terminated. There won't be a warning and the debugger will not show something clearly wrong, you will just be repeating what happens to be the last letter, and that is unlikely to give you the right suffix-order.